File: mimeparse.erl
%%% @author Steve Vinoski <vinoski@ieee.org> [http://steve.vinoski.net/]
%%% @doc MIME-Type Parser based on Joe Gregorio's mimeparse.py Python module. This module
%%%      provides basic functions for handling mime-types. It can handle matching mime-types
%%%      against a list of media-ranges. Comments are mostly excerpted from the original.
%%% @reference See <a href="http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.1">
%%%            RFC 2616, section 14.1</a> for a complete explanation of mime-type handling.
%%% @reference <a href="http://code.google.com/p/mimeparse/">mimeparse</a>

-module(mimeparse).
-author('vinoski@ieee.org').
-export([parse_mime_type/1, parse_media_range/1, quality/2, best_match/2]).
-export([test/0]).

%%% @type mime_type() = string().
%%% @type media_range() = mime_type().
%%% @type type() = atom().
%%% @type subtype() = atom().
%%% @type value() = string().
%%% @type param() = {atom(), value()}.
%%% @type mime_tuple() = {type(), subtype(), [param()]}.
%%% @type parsed_ranges() = [mime_tuple()].


%% @spec parse_mime_type(Mime_type::mime_type()) -> mime_tuple()
%% @doc Parses a mime-type into its component parts.
%%      Returns a tuple of the {type, subtype, params} where 'params' is a proplist
%%      of all the parameters for the mime-type. For example, the mime-type
%%      "application/xhtml;q=0.5" would get parsed into:
%%
%%         {application, xhtml, [{q, "0.5"}]}
%%
parse_mime_type(Mime_type) ->
    [Full_type | Parts] = string:tokens(string:strip(Mime_type), ";"),
    Params = lists:map(fun({K}) ->
                               {list_to_atom(K), ""};
                          ({K, V}) ->
                               {list_to_atom(K), V}
                       end, [list_to_tuple([string:strip(S) || S <- string:tokens(Param, "=")])
                             || Param <- Parts]),
    [Type, Subtype] = case Full_type of
                          % Java URLConnection class sends an Accept header that includes a single "*"
                          % Turn it into a legal wildcard.
                          "*" ->
                              ['*', '*'];
                          _ ->
                              [list_to_atom(string:strip(S)) || S <- string:tokens(Full_type, "/")]
                      end,
    {Type, Subtype, Params}.

%% @spec parse_media_range(Range::media_range()) -> mime_tuple()
%% @doc Parses a media-range into its component parts.
%%      Media-ranges are mime-types with wildcards and a 'q' quality parameter. This
%%      function performs the same as parse_mime_type/1 except that it also
%%      guarantees that there is a value for 'q' in the params dictionary, filling it
%%      in with a proper default if necessary.
%% @see parse_mime_type/1
%%
parse_media_range(Range) ->
    {Type, Subtype, Params} = parse_mime_type(Range),
    Default_q = {q, "1"},
    New_q = case lists:keysearch(q, 1, Params) of
                false ->
                    Default_q;
                {value, {q, Value}} ->
                    New_value = case Value of
                                    [$.|_] ->
                                        string:concat("0", Value);
                                    _ ->
                                        Value
                                end,
                    case number_to_float(New_value) of
                        {error, _} ->
                            Default_q;
                        Float ->
                            if
                                Float > 1.0;
                                Float < 0.0 ->
                                    Default_q;
                                true ->
                                    {q, New_value}
                            end
                    end
            end,
    {Type, Subtype, lists:keystore(q, 1, Params, New_q)}.

%% @spec quality(Mime_type::mime_type(), Range::media_range()) -> float()
%% @doc Determines the quality ('q') of a mime-type when compared against a media-range.
%%      Returns the quality 'q' of a mime_type when compared against the media-ranges
%%      in ranges. For example, given this media-range:
%%
%%      "text/*;q=0.3, text/html;q=0.7, text/html;level=1, text/html;level=2;q=0.4, */*;q=0.5"
%%
%%      this function returns 0.7.
%%
quality(Mime_type, Range) ->
    {_, Q} = quality_parsed(Mime_type, [parse_media_range(R) || R <- string:tokens(Range, ",")]),
    Q.

%% @spec best_match(Supported_mime_types::[mime_type()], Header::media_range()) -> mime_type()
%% @doc Choose the mime-type with the highest quality ('q') from a list of candidates.
%%      Takes a list of supported mime-types and finds the best match for all the
%%      media-ranges listed in Header. The value of Header must be a string that
%%      conforms to the format of the HTTP Accept: header. The value of
%%      Supported_mime_types is a list of mime-types.
%%
%%       best_match(["application/xbel+xml", "text/xml"], "text/*;q=0.5,*/*; q=0.1").
%%
%%      returns "text/xml".
%%
best_match(Supported_mime_types, Header) ->
    Parsed_header = [parse_media_range(R) || R <- string:tokens(Header, ",")],
    Best_match = lists:foldl(fun(Mime_type, {{Best_fit, Best_q}, _}=Best_so_far) ->
                                     Score = {Fitness, Q} = quality_parsed(Mime_type, Parsed_header),
                                     if
                                         Fitness >= Best_fit andalso Q >= Best_q ->
                                             {Score, Mime_type};
                                         true ->
                                             Best_so_far
                                     end
                             end, {{-1, 0.0}, ""}, Supported_mime_types),
    case Best_match of
        {{_, 0.0}, _} ->
            "";
        {_Score, Mime_type} ->
            Mime_type
    end.

%% Internal functions.

quality_parsed(Mime_type, Parsed_ranges) ->
    {Target_type, Target_subtype, Target_params} = parse_media_range(Mime_type),
    lists:foldl(
      fun({Type, Subtype, Params}, {Best_fitness, _}=Best_so_far) ->
              case possible_best_fit(Target_type, Type, Target_subtype, Subtype) of
                  true ->
                      Type_fitness = type_fitness(Target_type, Type, 100),
                      Subtype_fitness = type_fitness(Target_subtype, Subtype, 10),
                      Param_matches = count_param_matches(Target_params, Params),
                      Fitness = Type_fitness + Subtype_fitness + Param_matches,
                      if
                          Fitness > Best_fitness ->
                              {Fitness, number_to_float(proplists:get_value(q, Params))};
                          true ->
                              Best_so_far
                      end;
                  _ ->
                      Best_so_far
              end
      end, {-1, 0.0}, Parsed_ranges).

count_param_matches(Target_params, Params) ->
    lists:foldl(fun({q,_}, Acc) ->
                        Acc;
                   ({Key,Value}, Acc) ->
                        case lists:keysearch(Key, 1, Params) of
                            false ->
                                Acc;
                            {value, {Key, Value}} ->
                                1 + Acc;
                            _ ->
                                Acc
                        end
                end, 0, Target_params).

possible_best_fit(Target_type, Type, Target_subtype, Subtype) ->
    (lists:member(Type, ['*', Target_type]) orelse Target_type =:= '*')
        andalso
          (lists:member(Subtype, ['*', Target_subtype]) orelse
           Target_subtype =:= '*').

type_fitness(Target_type, Type, Value) ->
    case Type of
        Target_type -> Value;
        _ -> 0
    end.

number_to_float(String) ->
    case string:to_float(String) of
        {error, _}=Error ->
            case string:to_integer(String) of
                {error, _} ->
                    Error;
                {Num, _} ->
                    float(Num)
            end;
        {Float, _} ->
            Float
    end.

%% Tests.

test_parse_media_range() ->
    {application, xml, [{q, "1"}]} = mimeparse:parse_media_range("application/xml;q=1"),
    {application, xml, [{q, "1"}]} = mimeparse:parse_media_range("application/xml"),
    {application, xml, [{q, "1"}]} = mimeparse:parse_media_range("application/xml;q="),
    {application, xml, [{q, "1"}]} = mimeparse:parse_media_range("application/xml ; q="),
    {application, xml, [{q, "1"}, {b, "other"}]} =
        mimeparse:parse_media_range("application/xml ; q=1;b=other"),
    {application, xml, [{q, "1"}, {b, "other"}]} =
        mimeparse:parse_media_range("application/xml ; q=2;b=other"),
    % Java URLConnection class sends an Accept header with a single *
    {'*', '*', [{q, "0.2"}]} = mimeparse:parse_media_range(" *; q=.2"),
    ok.

test_rfc_2616_example() ->
    Accept = "text/*;q=0.3, text/html;q=0.7, text/html;level=1, text/html;level=2;q=0.4, */*;q=0.5",
    1.0 = mimeparse:quality("text/html;level=1", Accept),
    0.7 = mimeparse:quality("text/html", Accept),
    0.3 = mimeparse:quality("text/plain", Accept),
    0.5 = mimeparse:quality("image/jpeg", Accept),
    0.4 = mimeparse:quality("text/html;level=2", Accept),
    0.7 = mimeparse:quality("text/html;level=3", Accept),
    ok.

test_best_match() ->
    Mime_types_supported1 = ["application/xbel+xml", "application/xml"],
    % direct match
    "application/xbel+xml" = mimeparse:best_match(Mime_types_supported1, "application/xbel+xml"),
    % direct match with a q parameter
    "application/xbel+xml" = mimeparse:best_match(Mime_types_supported1, "application/xbel+xml; q=1"),
    % direct match of our second choice with a q parameter
    "application/xml" = mimeparse:best_match(Mime_types_supported1, "application/xml; q=1"),
    % match using a subtype wildcard
    "application/xml" = mimeparse:best_match(Mime_types_supported1, "application/*; q=1"),
    % match using a type wildcard
    "application/xml" = mimeparse:best_match(Mime_types_supported1, "*/*"),

    Mime_types_supported2 = ["application/xbel+xml", "text/xml"],
    % match using a type versus a lower weighted subtype
    "text/xml" = mimeparse:best_match(Mime_types_supported2, "text/*;q=0.5,*/*; q=0.1"),
    % fail to match anything
    "" = mimeparse:best_match(Mime_types_supported2, "text/html,application/atom+xml; q=0.9"),

    % common AJAX scenario
    Mime_types_supported3 = ["application/json", "text/html"],
    "application/json" = mimeparse:best_match(Mime_types_supported3, "application/json, text/javascript, */*"),
    % verify fitness ordering
    "application/json" = mimeparse:best_match(Mime_types_supported3, "application/json, text/html;q=0.9"),

    ok.

test_support_wildcards() ->
    Mime_types_supported = ["image/*", "application/xml"],
    % match using a type wildcard
    "image/*" = mimeparse:best_match(Mime_types_supported, "image/png"),
    % match using a wildcard for both requested and supported
    "image/*" = mimeparse:best_match(Mime_types_supported, "image/*"),
    ok.

test() ->
    test_parse_media_range(),
    test_rfc_2616_example(),
    test_best_match(),
    test_support_wildcards().

File: mimeparse.js
// mimeparse.js
// 
// This module provides basic functions for handling mime-types. It can
// handle matching mime-types against a list of media-ranges. See section
// 14.1 of the HTTP specification [RFC 2616] for a complete explanation.
//
//   http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.1
// 
// A port to JavaScript of Joe Gregorio's MIME-Type Parser:
// 
//   http://code.google.com/p/mimeparse/
// 
// Ported by J. Chris Anderson <jchris@apache.org>, targeting the Spidermonkey runtime.
// 
// To run the tests, open mimeparse-js-test.html in a browser.
// Ported from version 0.1.2
// Comments are mostly excerpted from the original. 

var Mimeparse = (function() {
  // private helpers
  function strip(string) {
    return string.replace(/^\s+/, '').replace(/\s+$/, '')
  };

  function parseRanges(ranges) {
    var parsedRanges = [], rangeParts = ranges.split(",");
    for (var i=0; i < rangeParts.length; i++) {
      parsedRanges.push(publicMethods.parseMediaRange(rangeParts[i]))
    };
    return parsedRanges;
  };

  var publicMethods = {
    // Carves up a mime-type and returns an Array of the
    //  [type, subtype, params] where "params" is a Hash of all
    //  the parameters for the media range.
    //  
    // For example, the media range "application/xhtml;q=0.5" would
    //  get parsed into:
    //  
    // ["application", "xhtml", { "q" : "0.5" }]
    parseMimeType : function(mimeType) {
      var fullType, typeParts, params = {}, parts = mimeType.split(';');
      for (var i=0; i < parts.length; i++) {
        var p = parts[i].split('=');
        if (p.length == 2) {
          params[strip(p[0])] = strip(p[1]);
        }
      };
      fullType = parts[0].replace(/^\s+/, '').replace(/\s+$/, '');
      if (fullType == '*') fullType = '*/*';
      typeParts = fullType.split('/');
      return [typeParts[0], typeParts[1], params];
    },

    // Carves up a media range and returns an Array of the
    //  [type, subtype, params] where "params" is a Object with
    //  all the parameters for the media range.
    //  
    // For example, the media range "application/*;q=0.5" would
    //  get parsed into:
    //  
    // ["application", "*", { "q" : "0.5" }]
    //  
    // In addition this function also guarantees that there
    //  is a value for "q" in the params dictionary, filling it
    //  in with a proper default if necessary.
    parseMediaRange : function(range) {
      var q, parsedType = this.parseMimeType(range);
      if (!parsedType[2]['q']) {
        parsedType[2]['q'] = '1';      
      } else {
        q = parseFloat(parsedType[2]['q']);
        if (isNaN(q)) {
          parsedType[2]['q'] = '1';      
        } else if (q > 1 || q < 0) {
          parsedType[2]['q'] = '1';      
        }
      }
      return parsedType;
    },

    // Find the best match for a given mime-type against 
    // a list of media_ranges that have already been 
    // parsed by parseMediaRange(). Returns an array of
    // the fitness value and the value of the 'q' quality
    // parameter of the best match, or (-1, 0) if no match
    // was found. Just as for qualityParsed(), 'parsed_ranges'
    // must be a list of parsed media ranges.
    fitnessAndQualityParsed : function(mimeType, parsedRanges) {
      var bestFitness = -1, bestFitQ = 0, target = this.parseMediaRange(mimeType);
      var targetType = target[0], targetSubtype = target[1], targetParams = target[2];
      
      for (var i=0; i < parsedRanges.length; i++) {
        var parsed = parsedRanges[i];
        var type = parsed[0], subtype = parsed[1], params = parsed[2];
        if ((type == targetType || type == "*" || targetType == "*") && 
          (subtype == targetSubtype || subtype == "*" || targetSubtype == "*")) {
          var matchCount = 0;
          for (param in targetParams) {
            if (param != 'q' && params[param] && params[param] == targetParams[param]) {
              matchCount += 1;
            }
          }
          
          var fitness = (type == targetType) ? 100 : 0;
          fitness += (subtype == targetSubtype) ? 10 : 0;
          fitness += matchCount;
          
          if (fitness > bestFitness) {
            bestFitness = fitness;
            bestFitQ = params["q"];
          }
        }
      };
      return [bestFitness, parseFloat(bestFitQ)];
    },

    // Find the best match for a given mime-type against
    // a list of media_ranges that have already been
    // parsed by parseMediaRange(). Returns the
    // 'q' quality parameter of the best match, 0 if no
    // match was found. This function bahaves the same as quality()
    // except that 'parsedRanges' must be a list of
    // parsed media ranges.
    qualityParsed : function(mimeType, parsedRanges) {
      return this.fitnessAndQualityParsed(mimeType, parsedRanges)[1];
    },
    
    // Returns the quality 'q' of a mime-type when compared
    // against the media-ranges in ranges. For example:
    // 
    // >>> Mimeparse.quality('text/html','text/*;q=0.3, text/html;q=0.7, text/html;level=1, text/html;level=2;q=0.4, */*;q=0.5')
    // 0.7
    quality : function(mimeType, ranges) {
      return this.qualityParsed(mimeType, parseRanges(ranges));
    },
    
    // Takes a list of supported mime-types and finds the best
    // match for all the media-ranges listed in header. The value of
    // header must be a string that conforms to the format of the 
    // HTTP Accept: header. The value of 'supported' is a list of
    // mime-types.
    // 
    // >>> bestMatch(['application/xbel+xml', 'text/xml'], 'text/*;q=0.5,*/*; q=0.1')
    // 'text/xml'
    bestMatch : function(supported, header) {
      var parsedHeader = parseRanges(header);
      var weighted = [];
      for (var i=0; i < supported.length; i++) {
        weighted.push([publicMethods.fitnessAndQualityParsed(supported[i], parsedHeader), supported[i]])
      };
      weighted.sort();
      // console.log(weighted)
      return weighted[weighted.length-1][0][1] ? weighted[weighted.length-1][1] : '';
    }
  }
  return publicMethods;
})();




Mimeparse.tests = {
  test_parseMediaRange : function() {
    T(equals(["application", "xml", {"q" : "1"}], 
      Mimeparse.parseMediaRange("application/xml;q=1")));
      
    T(equals(["application", "xml", {"q" : "1"}], 
      Mimeparse.parseMediaRange("application/xml")));
      
    T(equals(["application", "xml", {"q" : "1"}], 
      Mimeparse.parseMediaRange("application/xml;q=")));

    T(equals(["application", "xml", {"q" : "1"}], 
      Mimeparse.parseMediaRange("application/xml ; q=")));
    
    T(equals(["application", "xml", {"q" : "1", "b" : "other"}], 
      Mimeparse.parseMediaRange("application/xml ; q=1;b=other")));

    T(equals(["application", "xml", {"q" : "1", "b" : "other"}], 
      Mimeparse.parseMediaRange("application/xml ; q=2;b=other")));

    // Java URLConnection class sends an Accept header that includes a single "*"
    T(equals(["*", "*", {"q" : ".2"}], 
      Mimeparse.parseMediaRange(" *; q=.2")));
  },
  test_rfc_2616_example : function() {
    var accept = "text/*;q=0.3, text/html;q=0.7, text/html;level=1, text/html;level=2;q=0.4, */*;q=0.5";

    T(equals(1, Mimeparse.quality("text/html;level=1", accept)));
    T(equals(0.7, Mimeparse.quality("text/html", accept)));
    T(equals(0.3, Mimeparse.quality("text/plain", accept)));
    T(equals(0.5, Mimeparse.quality("image/jpeg", accept)));
    T(equals(0.4, Mimeparse.quality("text/html;level=2", accept)));
    T(equals(0.7, Mimeparse.quality("text/html;level=3", accept))); 
  },
  test_bestMatch : function() {
    var mimeTypesSupported = ['application/xbel+xml', 'application/xml'];
    // direct match
    T(equals(Mimeparse.bestMatch(mimeTypesSupported, 'application/xbel+xml'), 'application/xbel+xml'));
    // direct match with a q parameter
    T(equals(Mimeparse.bestMatch(mimeTypesSupported, 'application/xbel+xml; q=1'), 'application/xbel+xml'));

    // direct match of our second choice with a q parameter
    T(equals(Mimeparse.bestMatch(mimeTypesSupported, 'application/xml; q=1'), 'application/xml'));
    // match using a subtype wildcard
    T(equals(Mimeparse.bestMatch(mimeTypesSupported, 'application/*; q=1'), 'application/xml'));
    // match using a type wildcard
    T(equals(Mimeparse.bestMatch(mimeTypesSupported, '*/*'), 'application/xml'));

    mimeTypesSupported = ['application/xbel+xml', 'text/xml']
    // match using a type versus a lower weighted subtype
    T(equals(Mimeparse.bestMatch(mimeTypesSupported, 'text/*;q=0.5,*/*; q=0.1'), 'text/xml'));
    // fail to match anything
    T(equals(Mimeparse.bestMatch(mimeTypesSupported, 'text/html,application/atom+xml; q=0.9'), ''));

    // common AJAX scenario
    mimeTypesSupported = ['application/json', 'text/html']
    T(equals(Mimeparse.bestMatch(mimeTypesSupported, 'application/json, text/javascript, */*'), 'application/json'));
    // verify fitness ordering
    T(equals(Mimeparse.bestMatch(mimeTypesSupported, 'application/json, text/html;q=0.9'), 'application/json'));
    
  },
  test_support_wildcards : function() {
    var mime_types_supported = ['image/*', 'application/xml']
    // match using a type wildcard
    T(equals(Mimeparse.bestMatch(mime_types_supported, 'image/png'), 'image/*'));
    // match using a wildcard for both requested and supported 
    T(equals(Mimeparse.bestMatch(mime_types_supported, 'image/*'), 'image/*'));
    
  }
}



Mimeparse.runTests = function(outputFun) {

  // from CouchDB's Test Runner (Apache 2.0 license)
  function patchTest(fun) {
    var source = fun.toString();
    var output = "";
    var i = 0;
    var testMarker = "T("
    while (i < source.length) {
      var testStart = source.indexOf(testMarker, i);
      if (testStart == -1) {
        output = output + source.substring(i, source.length);
        break;
      }
      var testEnd = source.indexOf(");", testStart);
      var testCode = source.substring(testStart + testMarker.length, testEnd);
      output += source.substring(i, testStart) + "T(" + testCode + "," + JSON.stringify(testCode);
      i = testEnd;
    }
    try {
      return eval("(" + output + ")");
    } catch (e) {
      return null;
    }
  }
  
  function T(arg1, arg2) {
    var message = (arg2 != null ? unescape(arg2) : arg1).toString();
    if (arg1) {
      outputFun('<strong style="color:#0d0;">OK:</strong> <tt>'+message+'</tt>');
    } else {
      // console.log(arg2)
      outputFun('<strong style="color:#d00;">FAIL: <tt>'+message+'</tt></strong>');
    }
  }
  
  function equals(a,b) {
    if (a === b) return true;
    try {
      // console.log('equals')
      // console.log(repr(a))
      // console.log(repr(b))
      return repr(a) === repr(b);
    } catch (e) {
      return false;
    }
  }

  function repr(val) {
    if (val === undefined) {
      return null;
    } else if (val === null) {
      return "null";
    } else {
      return JSON.stringify(val);
    }
  }
  
  outputFun("Starting tests.");
  for (test in Mimeparse.tests) {
    outputFun("Running <tt>"+test+"</tt>");
    var testFun = Mimeparse.tests[test];
    try {
      testFun = patchTest(testFun) || testFun;
      testFun();
    } catch (e) {
      outputFun('<strong style="color:#f00;">ERROR: '+e.toString()+'</strong>');
      // console.log(e)
    }
  }
  outputFun("Finished tests.");
  
};
